bach/bach/partitioning.py

```diff
@@ -175,8 +180,14 @@ class GroupingSet(GroupingList):
     Abstraction to support SQLs
     GROUP BY GROUPING SETS ((colA,colB),(ColA),(ColC))
     """
-    def get_group_by_column_expression(self):
-        grouping_expr_list = [g.get_group_by_column_expression() for g in self._grouping_list]
+    def get_group_by_column_expression(self) -> Optional[Expression]:
+        grouping_optional_expr_list = [g.get_group_by_column_expression() for g in self._grouping_list]
+
+        # ensure that there are no None expressions in the list. If there are that's a bug, and we are
```

Comment:
mypy complained about the `Expression.construct` call below. Previously this function wasn't checked. 

This shouldn't break if it wasn't already broken, but I am not 100% sure whether the assert will always hold

Reply:
Why is it typed as `Optional`? and also in [`GroupBy.get_group_by_column_expression`](https://github.com/objectiv/objectiv-analytics/blob/b7615e4702a4e86b4c1d3b76129161bf67174e50/bach/bach/partitioning.py#L93-L99)? If I'm not wrong, that function will always return an Expression.

Reply:
anyways, I think it's better to do smth like this
```suggestion
    def get_group_by_column_expression(self) -> Expression:
        grouping_expr_list = []
        for g in self._grouping_list:
            expr = g.get_group_by_column_expression()
            if expr:
                grouping_expr_list.append(expr)

        if not grouping_expr_list:
            raise Exception('no valid group by expressions.')

        fmtstr = ', '.join(["{}"] * len(grouping_expr_list))
        return Expression.construct(fmtstr, *grouping_expr_list)
```

Reply:
Problem I had is that I don't really understand this code, and reasoning through it all would take a lot of time, so I figured I create a PR, and hopefully one of you would know if we can get `None` here and if that happens what to do. And if nobody does we can decide to either invest time or keep the code just as broken/well working as it was.

I think the new code breaks in exactly the same cases that the old code did, so it's no worse or better. Just mypy is happy now. Although you are probably right that we should raise a real excception, and not rely on `assert` as that is sometimes disabled.

Regarding your solution: I am not sure if you solution is correct, maybe if we do get a `None` here that is a bug, and if we just ignore it we might get wrong data? Again: I don't really understand this code, so you might well be right, just wondering whether you know what this does.

Reply:
@jansentom perhaps you have some insights here too?

Reply:
I'd be happy with raising an exception, this should not happen.

Reply:
This code is not very well tested and it's been a while since I wrote it. I suggest we just throw the exception and leave it as is.

Reply:
Ok, now it raises a regular Exception in the case of a None